src/EditorFeatures/Core/Implementation/Classification/SemanticClassificationUtilities.cs

```diff
@@ -22,6 +27,17 @@ namespace Microsoft.CodeAnalysis.Editor.Implementation.Classification
 {
     internal static class SemanticClassificationUtilities
     {
+        /// <summary>
+        /// Mapping from workspaces to a task representing when they are fully loaded.  While this task is not complete,
+        /// the workspace is still loading.  Once complete the workspace is loaded.  We store the task around, instead
+        /// of just awaiting <see cref="IWorkspaceStatusService.IsFullyLoadedAsync"/> as actually awaiting that call
+        /// takes non-neglible time (upwards of several hundred ms, to a second), whereas we can actually just use the
+        /// status of the <see cref="IWorkspaceStatusService.WaitUntilFullyLoadedAsync"/> task to know when we have
+        /// actually transitioned to a loaded state.
+        /// </summary>
+        private static readonly ConditionalWeakTable<Workspace, Task> s_workspaceToFullyLoadedStateTask =
```

Comment:
Should we just make the IWorkspaceStatusService cache the task rather than making classification specifically have to do the caching? Feels like a bit of a leaky abstraction here. (Consider if we make a _second_ feature do this pattern...)

Reply:
This can be handled separately but looking at the service it's not clear to me why we wouldn't just have it internally cache a task and then just hand those out to callers for WaitUntilFullyLoadedAsync, but maybe there's some subletly I'm missing.

Reply:
i'm ok doing that.

Reply:
i tried goign down this path but backed off. core issue is that the main service heavily uses cancellation tokens, and it's harder to cache tasks with that model if we want cancellation to truly happen.  In the code here, i effectively do not cancel as it's a one-and-done thing where i just look at hte completed state.  i didn't feel comfortable greatly changing the lower level system here.